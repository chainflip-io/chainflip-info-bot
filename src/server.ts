import { createBullBoard } from '@bull-board/api';
import { BullMQAdapter } from '@bull-board/api/bullMQAdapter.js';
import { FastifyAdapter } from '@bull-board/fastify';
import { type Queue } from 'bullmq';
import fastify from 'fastify';
import { GraphQLClient } from 'graphql-request';
import env from './env.js';
import logger from './utils/logger.js';

const app = fastify({ loggerInstance: logger, disableRequestLogging: true });

export const explorerClient = new GraphQLClient(env.EXPLORER_GATEWAY_URL);
export const lpClient = new GraphQLClient(env.LP_GATEWAY_URL);

export const createServer = (queues: Queue[]) => {
  const serverAdapter = new FastifyAdapter();

  createBullBoard({
    queues: queues.map((q) => new BullMQAdapter(q)),
    serverAdapter,
  });

  const basePath = '/admin/queues';
  serverAdapter.setBasePath(basePath);
  app.register(serverAdapter.registerPlugin(), { prefix: basePath, basePath: '/' });

  app.get('/health', () => ({ status: 'ok' }));

  app.get('/', (req, res) => res.redirect('/admin/queues', 302));

  return app;
};
