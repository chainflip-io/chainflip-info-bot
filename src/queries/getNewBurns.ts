import assert from 'assert';
import { BigNumber } from 'bignumber.js';
import { FLIP_DECIMAL_POINTS } from '../consts.js';
import { gql } from '../graphql/generated/gql.js';
import { explorerClient } from '../server.js';

const getNewBurnsQuery = gql(/* GraphQL */ `
  query getNewBurns($id: Int!) {
    burns: allBurns(filter: { id: { greaterThan: $id } }, orderBy: ID_DESC) {
      nodes {
        id
        amount
        valueUsd
        event: eventByEventId {
          blockId
          indexInBlock
          block: blockByBlockId {
            timestamp
          }
        }
      }
    }
  }
`);

export default async function getNewBurns(latestBurnId: number) {
  const result = await explorerClient.request(getNewBurnsQuery, { id: latestBurnId });

  assert(result.burns, 'burns is required');

  return result.burns.nodes.map((node) => ({
    ...node,
    amount: new BigNumber(node.amount).shiftedBy(-FLIP_DECIMAL_POINTS),
  }));
}
