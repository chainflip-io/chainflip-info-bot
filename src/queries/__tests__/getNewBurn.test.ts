import { describe, expect, it, vi } from 'vitest';
import { explorerClient } from '../../server.js';
import getNewBurns from '../getNewBurn.js';

describe('getNewBurns', () => {
  it('should return the latest burn id', async () => {
    vi.mocked(explorerClient.request).mockResolvedValue({
      burns: {
        nodes: [
          {
            id: 11,
            amount: '123456789012345',
            valueUsd: '4.21',
            event: {
              blockId: 1,
              indexInBlock: 1,
              block: {
                timestamp: '2024-10-31T15:15:00.000Z',
              },
            },
          },
        ],
      },
    });

    expect(await getNewBurns(10)).toMatchInlineSnapshot(`
      {
        "amount": "0.000123456789012345",
        "event": {
          "block": {
            "timestamp": "2024-10-31T15:15:00.000Z",
          },
          "blockId": 1,
          "indexInBlock": 1,
        },
        "id": 11,
        "valueUsd": "4.21",
      }
    `);

    expect(explorerClient.request).toHaveBeenCalledTimes(1);
  });
});
