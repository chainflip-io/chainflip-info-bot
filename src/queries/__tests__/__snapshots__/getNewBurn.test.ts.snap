// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getNewBurns > should return the latest burn id 1`] = `
{
  "amount": "0.000123456789012345",
  "event": {
    "block": {
      "timestamp": "2024-10-31T15:15:00.000Z",
    },
    "blockId": 1,
    "indexInBlock": 1,
  },
  "id": 11,
  "valueUsd": "4.21",
}
`;
